tactics/autorewrite.ml

```diff
@@ -408,32 +407,46 @@ let gen_auto_multi_rewrite conds tac_main lbas cl =
   let try_do_hyps treat_id l =
     autorewrite_multi_in ~conds (List.map treat_id l) tac_main lbas
   in
+  let concl_tac = (if cl.concl_occs != NoOccurrences then autorewrite ~conds tac_main lbas else Proofview.tclUNIT ()) in
   if not (Locusops.is_all_occurrences cl.concl_occs) &&
      cl.concl_occs != NoOccurrences
   then
     let info = Exninfo.reify () in
     Tacticals.tclZEROMSG ~info (str"The \"at\" syntax isn't available yet for the autorewrite tactic.")
   else
-    let compose_tac t1 t2 =
-      match cl.onhyps with
-        | Some [] -> t1
-        | _ ->      Tacticals.tclTHENFIRST t1 t2
-    in
-    compose_tac
-        (if cl.concl_occs != NoOccurrences then autorewrite ~conds tac_main lbas else Proofview.tclUNIT ())
-        (match cl.onhyps with
-           | Some l -> try_do_hyps (fun ((_,id),_) -> id) l
-           | None ->
-                 (* try to rewrite in all hypothesis
-                    (except maybe the rewritten one) *)
-               Proofview.Goal.enter begin fun gl ->
-                 let ids = Tacmach.pf_ids_of_hyps gl in
-                 try_do_hyps (fun id -> id)  ids
-               end)
+    match cl.onhyps with
+    | Some [] -> concl_tac
+    | Some l -> Tacticals.tclTHENFIRST concl_tac (try_do_hyps (fun ((_,id),_) -> id) l)
+    | None ->
+      let hyp_tac =
+        (* try to rewrite in all hypothesis (except maybe the rewritten one) *)
+        Proofview.Goal.enter begin fun gl ->
+          let ids = Tacmach.pf_ids_of_hyps gl in
+          try_do_hyps (fun id -> id)  ids
+        end
+      in
+      Tacticals.tclTHENFIRST concl_tac hyp_tac
 
 let auto_multi_rewrite ?(conds=Naive) lems cl =
   Proofview.wrap_exceptions (fun () -> gen_auto_multi_rewrite conds (Proofview.tclUNIT()) lems cl)
 
+(* Same hack as auto hints: we generate an essentially unique identifier for
+   rewrite hints. *)
```

Comment:
Should we expose this somewhere (in Lib?) instead of copying it around?

Reply:
This should go away at some point. The only reason we need this is the infamous ability to add non-globref hint entries.